sdlib/d/gc/tcache.d

```diff
@@ -78,23 +160,37 @@ public:
 				copySize = getSizeFromClass(oldSizeClass);
 			}
 		} else {
-			auto esize = pd.extent.size;
-			if (alignUp(size, PageSize) == esize) {
+			oldFill = pd.extent.allocSize;
+
+			// Prohibit resize below appendable fill:
+			if (size < oldFill)
+				return ptr;
+
+			auto oldSize = oldFill ? oldFill : pd.extent.size;
+
+			if (alignUp(size, PageSize) == alignUp(oldSize, PageSize)) {
 				return ptr;
 			}
 
 			// TODO: Try to extend/shrink in place.
-			import d.gc.util;
-			copySize = min(size, esize);
+			copySize = min(size, oldSize);
 		}
 
 		containsPointers = (containsPointers | pd.containsPointers) != 0;
-		auto newPtr = alloc(size, containsPointers);
+		auto useSize = oldFill ? upsizeToLarge(size) : size;
```

Comment:
Here you recompute the size, so the previous short circuit is incorrect. This will also lead to quadratic behavior as the size will only be bumped up one page at the time instead of exponentially, which is linear.